internal/freelist/freelist_test.go

```diff
@@ -266,6 +356,250 @@ func TestFreelist_write(t *testing.T) {
+	require.Equal(t, common.Pgid(0), f.Allocate(common.Txid(4), 1))
+}
+
+func TestFreelist_E2E_MultiSpanOverflows(t *testing.T) {
+	f := newTestFreelist()
+	f.Init([]common.Pgid{})
+	f.Free(common.Txid(10), common.NewPage(20, common.LeafPageFlag, 0, 1))
+	f.Free(common.Txid(10), common.NewPage(25, common.LeafPageFlag, 0, 2))
+	f.Free(common.Txid(10), common.NewPage(35, common.LeafPageFlag, 0, 3))
+	f.Free(common.Txid(10), common.NewPage(39, common.LeafPageFlag, 0, 2))
+	f.Free(common.Txid(10), common.NewPage(45, common.LeafPageFlag, 0, 4))
+	requirePages(t, f, common.Pgids{}, common.Pgids{20, 21, 25, 26, 27, 35, 36, 37, 38, 39, 40, 41, 45, 46, 47, 48, 49})
+	f.ReleasePendingPages()
+	requirePages(t, f, common.Pgids{20, 21, 25, 26, 27, 35, 36, 37, 38, 39, 40, 41, 45, 46, 47, 48, 49}, common.Pgids{})
+
+	// that sequence, regardless of implementation, should always yield the same blocks of pages
+	allocSequence := []int{7, 5, 3, 2}
+	expectedSpanStarts := []common.Pgid{35, 45, 25, 20}
+	for i, pageNums := range allocSequence {
+		allocated := f.Allocate(common.Txid(11), pageNums)
+		require.Equal(t, expectedSpanStarts[i], allocated)
+		// ensure all pages in that span are not considered free anymore
+		for i := 0; i < pageNums; i++ {
+			require.False(t, f.Freed(allocated+common.Pgid(i)))
+		}
+	}
+}
+
+func TestFreelist_E2E_Rollbacks(t *testing.T) {
+	freelist := newTestFreelist()
+	freelist.Init([]common.Pgid{})
+	freelist.Free(common.Txid(2), common.NewPage(5, common.LeafPageFlag, 0, 1))
+	freelist.Free(common.Txid(2), common.NewPage(8, common.LeafPageFlag, 0, 0))
+	requirePages(t, freelist, common.Pgids{}, common.Pgids{5, 6, 8})
+	freelist.Rollback(common.Txid(2))
+	requirePages(t, freelist, common.Pgids{}, common.Pgids{})
+
+	// unknown transaction should not trigger anything
+	freelist.Free(common.Txid(4), common.NewPage(13, common.LeafPageFlag, 0, 3))
+	requirePages(t, freelist, common.Pgids{}, common.Pgids{13, 14, 15, 16})
+	freelist.ReleasePendingPages()
+	requirePages(t, freelist, common.Pgids{13, 14, 15, 16}, common.Pgids{})
+	freelist.Rollback(common.Txid(1337))
+	requirePages(t, freelist, common.Pgids{13, 14, 15, 16}, common.Pgids{})
+}
+
+func TestFreelist_E2E_RollbackPanics(t *testing.T) {
+	freelist := newTestFreelist()
+	freelist.Init([]common.Pgid{5})
+	requirePages(t, freelist, common.Pgids{5}, common.Pgids{})
+
+	_ = freelist.Allocate(common.Txid(5), 1)
+	require.Panics(t, func() {
+		// depending on the verification level, either should panic
+		freelist.Free(common.Txid(5), common.NewPage(5, common.LeafPageFlag, 0, 0))
+		freelist.Rollback(5)
+	})
+}
+
+func TestFreelist_E2E_ReadOnlyTxTracking(t *testing.T) {
+	freelist := newTestFreelist()
+	freelist.Init([]common.Pgid{})
+	freelist.Free(common.Txid(10), common.NewPage(10, common.LeafPageFlag, 0, 2))
+	freelist.Free(common.Txid(11), common.NewPage(20, common.LeafPageFlag, 0, 2))
+	freelist.Free(common.Txid(12), common.NewPage(30, common.LeafPageFlag, 0, 2))
```

Comment:
This doesn't make sense. Only one RW TXN is allowed at a time. The three `Free` calls are for three different TXID, it means there are 3 concurrent RW TXNs.

Reply:
it's a simple test setup, we can sprinkle the free calls around the `AddReadOnlyTXID` if this is more sensible and readable for you.

Reply:
I think we need to satisfy the following conditions,
- Only one RW TXN is allowed at a time. So yon can't call `Free` with a higher txid before you commit the previous RW TXN.
- When you create a readonly TXN, its TXID should equal to the latest committed RW TXN's txid.

Reply:
shall we add an assertion for this in a separate PR?

Reply:
> shall we add an assertion for this in a separate PR?

It's hard to add such assertion/verification in freelist layer, because the semantics is guaranteed by upper layer (bbolt) instead of freelist.  It's also an indication that the freelist API isn't well designed.

Reply:
Can you move this test into a separate PR so that we can merge this PR? thx

Reply:
yep, will do. brb

Reply:
removed